### DISCUSSION TITLE
How to choose the right strategy for building a Consul service? Consul DevOps or Ocelot C# Dev?

### DISCUSSION THREAD

**Discussion by RainySeason673** [ThreeMammals/Ocelot#2148] [Q&A] [answered]
## Expected Behavior / New Feature
hope add some configuration parameter of ocelot.consul :  use for choose consul service'name or service's address  when forward request to consul

## Actual Behavior / Motivation for New Feature
this is response error msg:
![image](https://github.com/user-attachments/assets/f8b55fe6-b18b-4021-907a-3637612d3ea8)

## Steps to Reproduce the Problem

  1.  my consul and microservice run at same CVM, services and nodes info:
![image](https://github.com/user-attachments/assets/b4d31fc3-b3f3-45c4-a5db-a82a6d144207)
![image](https://github.com/user-attachments/assets/cd7630a8-0d42-454a-892a-a1bca82ab6b5)

  2.  when ocelot forward the request to Consul, then will response error  like this:
![image](https://github.com/user-attachments/assets/103a9cfc-effe-45e2-a96e-18b4750972a8)

 3.  i can't debug invoke process, but i guess the [entry?.Node] is not null by response error
![image](https://github.com/user-attachments/assets/d7c3b1a9-e294-4b8f-bf6d-447f605ff8c7)

 4.  as you see, if  GetDownstreamHost return ip rather than name, this problem fixed. so, if can choose the  DownstreamHost by exposed configuration parameter , this problem will fix 

**thanks for all your replies!**

## Specifications

  - Version: 23.3.3
  - Platform: Tencent CVM
  - Subsystem: centos 8.5

**Comment by raman-m** [comment, marked as answer]
Hello Tang!

> ## Expected Behavior / New Feature
>hope add some configuration parameter of ocelot.consul :  use for choose consul service'name or service's address  when forward request to consul

Requests are routed to downstream services, not to Consul. **We will not be adding any JSON parameters to the `GlobalConfiguration.ServiceDiscoveryProvider` or at the route level.** One should act as a DevOps engineer by defining the correct node names, or as a C# developer by overriding the default service builder.

Have you read [Consul Service Builder](https://ocelot.readthedocs.io/en/latest/features/servicediscovery.html#consul-service-builder) docs?
The documents outline the following [best practice](https://ocelot.readthedocs.io/en/latest/features/servicediscovery.html#consul-service-builder:~:text=Some%20DevOps%20engineers%20choose%20to%20ignore%20node%20names%2C%20opting%20instead%20for%20abstract%20identifiers%20rather%20than%20actual%20hostnames.%20Our%20team%2C%20however%2C%20advocates%20for%20the%20assignment%20of%20real%20hostnames%20or%20IP%20addresses%20to%20node%20names%2C%20upholding%20this%20as%20a%20best%20practice.):
> Some DevOps engineers choose to ignore node names, opting instead for abstract identifiers rather than actual hostnames. Our team, however, advocates for the assignment of real hostnames or IP addresses to node names, upholding this as a best practice.

Ultimately, the decision comes down to selecting your role: Consul DevOps or Ocelot C# Developer?

**Comment by raman-m** [comment]
>   2.  when ocelot forward the request to Consul, then will response error  like this:
> ![image](https://github.com/user-attachments/assets/103a9cfc-effe-45e2-a96e-18b4750972a8)
> 

So, error is `ConnectionToDownstreamServiceError`. Exception is `HttpRequestException` with message text:
`Name or service not known (cluster-server1:31001)`
So, default Ocelot's service builder constructed your downstream host as `cluster-server1:31001` where `cluster-server1` is host, and 31001 is the port. Because `cluster-server1` is not valid DNS name, so you have this network connectivity error.

You need to consider the following 2 possible solutions:
- As a Consul DevOps engineer, you need to convert node names to real DNS hosts. And you must setup DNS system somehow.
- As a Ocelot C# developer, you have to use [AddConsul{T} method](https://ocelot.readthedocs.io/en/latest/features/servicediscovery.html#addconsul-t-method), and finally override the `GetDownstreamHost` virtual method like this: `protected override string GetDownstreamHost(ServiceEntry entry, Node node) => entry.Service.Address;`
  Bingo! 💥

**Comment by raman-m** [comment]
> 4. as you see, if GetDownstreamHost return ip rather than name, this problem fixed.

It seems you haven't read [the docs](https://ocelot.readthedocs.io/en/latest/features/servicediscovery.html#consul-service-builder), dude... This is definitely the fix, as it is already explained in the docs!

> so, if can choose the DownstreamHost by exposed configuration parameter , this problem will fix

I'm not sure which "configuration parameter" you're referring to. If you're suggesting the config parameter in `GlobalConfiguration.ServiceDiscoveryProvider` JSON, we consider this method to be excessive. We have opted against using JSON settings in favor of overriding C# virtual methods or adhering to Consul Node Name best practices.

Finally, I don't understand why you reported this issue. We are not going to provide any JSON parameters in the global `ServiceDiscoveryProvider` JSON at all! The reason is that it's impossible to describe the variety of Consul setups based on JSON; a more productive approach is to follow best practices or to override behavior in C#.

**Comment by RainySeason673** [comment]
Thank you for your explanation. I did not pay attention to this part of the document. I now know how to fix this problem as a best practice!

**Reply by raman-m** [reply]
Congratulations! 🥳
